Make RunCsc/Vbc scripts respect DOTNET_HOST_PATH

khyperia opened the pull request:
This is related to https://github.com/dotnet/cli/pull/7311

Logic: If `DOTNET_HOST_PATH` is set, use that as the path to `dotnet`. Otherwise, search in `PATH`.

agocke commented on build/NuGetAdditionalFiles/RunCsc:
```diff
@@ -1,5 +1,15 @@
+#  If VAR is non-empty => TEXT
+# Do this expression to make sure HOST_PATH ends with a slash,
+#  but if it's unset, have it be completely empty (look up dotnet in PATH)
+HOST_PATH=${DOTNET_HOST_PATH:+${DOTNET_HOST_PATH-}/}
```
Fancy :)

jaredpar replied:
Why would we do this vs. the approach in the CMD files: if `DOTNET_HOST_PATH` is defined then just use it directly?

khyperia replied:
It *is* doing that. `@{VAR:+TXT}` is equivalent to `if (defined VAR) then (TXT) else ("")` (pseudo-syntax).

The reason why it's not just directly `${DOTNET_HOST_PATH}dotnet` is: what if DOTNET_HOST_PATH doesn't end in a slash? Say, `/some/path`, we'd then be invoking `/some/pathdotnet`.

If we wrote `${DOTNET_HOST_PATH}/dotnet`, then an empty host path would cause `/dotnet` to be invoked, not `dotnet`.

The reason it's not an if statement is because bash doesn't have a "is this variable defined" builtin - the way to check if a var is defined is to use one of these features like `:+`, so if we're doing that anyway, why not just jump straight to the right thing.

khyperia replied:
Also, the reason there's a dash at the end of `${DOTNET_HOST_PATH-}` in the inner expression is because I wasn't sure of evaluation order (outer->inner or inner->outer). If the inner is evaluated first, and `DOTNET_HOST_PATH` is unset, then that would trigger an error (if you have sane bash options set). However, as we *know* it can possibly be unset in non-error-cases, adding the `-` does a simple "replace with empty string if unset" instead of erroring. (`${VAR-TXT}`/`${VAR:-TXT}` are "if `VAR` set, return `${VAR}`, else return `TXT`, with `:` behavior differing on if `VAR` is the empty string - when the replacement string *is* the empty string, though, it doesn't matter, so `-` is fine)

tl;dr variables are complicated

jaredpar replied:
Hmm, think I need to get a "bash for recovering unix guys" book. :)

jaredpar commented on build/NuGetAdditionalFiles/RunCsc.cmd:
```diff
@@ -1,2 +1,4 @@
 echo off
-dotnet %~dp0\csc.dll %*
+if defined DOTNET_HOST_PATH set HOST_PATH=%DOTNET_HOST_PATH%\
+if not defined HOST_PATH set HOST_PATH=
```
This line seems redundant. Why do we need it?

khyperia replied:
I wasn't sure how cmd treats unset variables vs. empty variables (e.g. if it raises an error in some cases if a variable is unset). I decided to not risk it and set it to the empty string if the variable is unset. If you know for sure that cmd always replaces unset variables with the empty string (without error/warning), I can remove this line.

khyperia replied:
I suppose this is actually kind of buggy, as it allows leakage of the `HOST_PATH` variable (if `DOTNET_HOST_PATH` is unset and `HOST_PATH` is set, it erroneously uses `HOST_PATH`). Changing it to `if not defined DOTNET_HOST_PATH set HOST_PATH=` would resolve that.

khyperia commented:
Ping @agocke, could you re-review the latest changes? Took feedback for the .cmd scripts (and fixed the issue I mentioned above). Ping @jaredpar to re-review and see if that addressed feedback.

Also, batch stuff is ridiculous - when in a REPL, doing `echo "%an_unset_var%"` prints `"%an_unset_var%"`, but doing *the same thing* in a .cmd script prints `""`. I can't find docs *anywhere*.

jaredpar commented on build/NuGetAdditionalFiles/RunCsc.cmd:
```diff
@@ -1,2 +1,7 @@
-echo off
-dotnet %~dp0\csc.dll %*
+@echo off
+if defined DOTNET_HOST_PATH (
```
You mentioned you researched CMD defined variables a bit. Did you come to have a preference between the following two styles we use?

``` cmd
if "%DOTNET_HOST_PATH%" != "" (
if defined DOTNET_HOST_PATH (
```

khyperia replied:
No - I'm not sure if there's a significant difference. `if defined` looks stylistically cleaner, though, and I'd be less nervous about it (because of the following).

I was specifically researching behavior of unset variables - for example, my comment earlier in this thread:

>  when in a REPL, doing `echo "%an_unset_var%"` prints `"%an_unset_var%"`, but doing *the same thing* in a .cmd script prints `""`.

I was trying to validate correct behavior, since it seems like CMD vars don't have two distinct null/empty states, only the one - so I was trying to validate that the `set HOST_PATH=` line (which unsets the var `HOST_PATH`) does the right thing in all scenarios (so we don't execute literally `%HOST_PATH%dotnet`), but all I'm really sure about at this point is that it works when testing locally :(
